### PR TITLE
Kgain e2e

### PR DESCRIPTION
**Pull request by JuergenSchreiber**
## Describe your changes
- added an end-to-end test for kgain determination inclusive the corresponding walker recipe
- compare the determined kgain and readnoise with the ones determined from TVAC

## Type of change

- New feature (non-breaking change which adds functionality)

## Checklist before requesting a review
- [x] I have linted my code
- [x] I have verified that all unit tests pass in a clean environment and added new unit tests, as needed
- [x] I have verified that all docstrings are properly formatted and added new documentation, as needed

### CODE REVIEW

**Comment by JuergenSchreiber** [comment]
For efficiency reasons I have only used a subset of the available TVAC noise data, so I had to reduce the n_cal and n_mean parameters to 3, I think I should change that back to default for the real e2e tests?

**Comment by semaphoreP** [comment]
What is the difference in efficiency? We ideally want to test against a realistic dataset, which includes assessing the computing resources we need. Note that we don't intend to run the e2e tests against every PR like we do for the unit tests.

File: corgidrp/recipe_templates/l1_to_kgain.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        },
+        {
+            "name" : "correct_nonlinearity",
+            "calibs" : {
+                "NonLinearityCalibration" : "AUTOMATIC"
+            }
+        },
+		{   "name" : "calibrate_kgain",
+		    "keywords" : {
+		        "n_cal" : 3,
```
Comment (semaphoreP) [inline corgidrp/recipe_templates/l1_to_kgain.json:30]: this is going to be the default recipe for the real data, so we should use the parameters we think are best for the real data.

File: tests/e2e_tests/kgain_e2e.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    for file in l1_data_filelist_same_exp:
+        image = data.Image(file)
+        # This should not be necessary anymore after the updates of the OBSTYPE keyword, up to now it is only "SCI"
+        if image.pri_hdr['OBSTYPE'] != 'MNFRAME':
+            image.pri_hdr['OBSTYPE'] = 'MNFRAME'
+            image.save(filename = file)
+        l1_data_filelist_range_exp.append(file)
+
+    walker.walk_corgidrp(l1_data_filelist_range_exp, "", kgain_outputdir, template="l1_to_kgain.json")
```
Comment (semaphoreP) [inline tests/e2e_tests/kgain_e2e.py:74]: Should we edit `walker.guess_template()` to handle how to process kgain without needing to pass into the template. Or can we not differentiate this data from the nonlinearity data?

File: tests/e2e_tests/kgain_e2e.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    print("determined read noise", new_readnoise)    
+    
+    diff_kgain = new_kgain - tvac_kgain
+    diff_readnoise = new_readnoise - tvac_readnoise
+    print ("difference to TVAC kgain:", diff_kgain)
+    print ("difference to TVAC read noise:", diff_readnoise)
+    print ("error of kgain:", kgain.error)
+    print ("error of  readnoise:", kgain.ext_hdr["RN_ERR"])
+
+    assert np.abs(diff_kgain) < 1
```
Comment (semaphoreP) [inline tests/e2e_tests/kgain_e2e.py:104]: are these differences small enough to be tolerated? We should try running on the full dataset and circle back to this.

**Review by semaphoreP — COMMENTED** [review]
Thanks for doing this Juergen. Fairly straightforward. We should definitely run the e2e on the full set of real data, and assess its performance. We need to also think about if there's a way to automatically identify this set of data (or is it degenerate with the nonlinearity recipe)?

File: corgidrp/recipe_templates/l1_to_kgain.json (hunk 1 above)
Comment (JuergenSchreiber) [reply corgidrp/recipe_templates/l1_to_kgain.json:30]: I leave it to the default values

File: tests/e2e_tests/kgain_e2e.py (hunk 3 above)
Comment (JuergenSchreiber) [reply tests/e2e_tests/kgain_e2e.py:104]: We could use the determined errors to define a tolerance.

Comment (JuergenSchreiber) [reply tests/e2e_tests/kgain_e2e.py:104]: Understood, so then just a very small number

File: corgidrp/recipe_templates/l1_to_kgain.json (hunk 1 above)
Comment (JuergenSchreiber) [reply corgidrp/recipe_templates/l1_to_kgain.json:30]: These values will be changed to default, that means they do not appear here anymore

**Comment by semaphoreP** [comment]
We recently learned that the Kgain and Nonlin algorithms were updated after the TVAC data products were reduced, and the version we ported over is the updated version. This means that the TVAC data products were generated with the old code unfortunately (this is what Max is referring to in his recent comments in #146). So we have to unfortunately do an extra step: run the TVAC data through the updated engineering pipeline and compare it with our ported version (for nonlin, Sergi confirmed this made the differences disappear). Can you do that? Hopefully, it won't take too long to get running. We can then store the output products on Box for this e2e test to compare against. 

Do you have a copy of the original Kgain code from the updated engineering pipeline? If not, @mygouf is working on releasing it, so you should contact her for the latest status.

**Comment by mygouf** [comment]
> We recently learned that the Kgain and Nonlin algorithms were updated after the TVAC data products were reduced, and the version we ported over is the updated version. This means that the TVAC data products were generated with the old code unfortunately (this is what Max is referring to in his recent comments in #146). So we have to unfortunately do an extra step: run the TVAC data through the updated engineering pipeline and compare it with our ported version (for nonlin, Sergi confirmed this made the differences disappear). Can you do that? Hopefully, it won't take too long to get running. We can then store the output products on Box for this e2e test to compare against.
> 
> Do you have a copy of the original Kgain code from the updated engineering pipeline? If not, @mygouf is working on releasing it, so you should contact her for the latest status.

kgain code is now released and available here: https://github.com/roman-corgi/cgi_iit_drp/tree/main/Calibration_NTR/cal/kgain

**Comment by JuergenSchreiber** [comment]
@mygouf, thanks. Do I have to take the same ROI as @hsergi did for nonlin e2e?

**Comment by maxwellmb** [comment]
> @mygouf, thanks. Do I have to take the same ROI as @hsergi did for nonlin e2e? OK, the ROI is defined in the kgain_params, but the offset_"xxxroi" is not applied.

I think you just want to reproduce as much as possible what the KGain TVAC code is doing, not necessarily what nonlin is doing, so you want to the right ROI from the TVAC KGain code.

**Comment by JuergenSchreiber** [comment]
OK, now it can be reviewed. The test results and performance I have put on confluence

**Review by maxwellmb — APPROVED** [review]
I'm going to approve the PR, but there are a couple oustanding issues for which I'm going to make issues: 

- As of right now, `walk_corgidrp` can't guess the template recipe, it must be passed in. We'd like it to be able to guess. 
- The results are not a perfect port to the engineering code. They are very close, but not perfect. We'd like to resolve this.

**Pull request merged**